### PR TITLE
feat(rust): streaming parquet from object_stores

### PR DESCRIPTION
**Pull request by winding-lines**
Supports reading parquet files with the object_store crate. Currently the `file` and `s3` protocols are linked in, gcs and azure can be added if this approach makes sense.

The only supported mode is `batched` since generally remote files are big. Setting a big chunk size will download all the file in one iteration.

This is my first PR in this repo, feedback appreciated.

### CODE REVIEW

**Comment by chitralverma** [comment]
@winding-lines AFAIK, `scan_parquet` and `read_parquet` can rely on `fsspec` filesystems to read data directly from S3, GCS and many more. so do we really need a separate `read_parquet_cloud` for this?

For a library point of view, it will be great for the users if they were flawlessly able to access their cloud data sets from the existing methods ?

**Comment by winding-lines** [comment]
> @winding-lines AFAIK, `scan_parquet` and `read_parquet` can rely on `fsspec` filesystems to read data directly from S3, GCS and many more. so do we really need a separate `read_parquet_cloud` for this?
> 
> For a library point of view, it will be great for the users if they were flawlessly able to access their cloud data sets from the existing methods ?

I personally believe that Python has achieved its amazing success in machine learning, data science and big data by providing great APIs on top of native libraries. You are right that fsspec is one such APIs, however I have concerns that it will impact performance in some use cases because of the way execution is serialized through Python. Some of these use cases:

1. I need to open hundreds of parquet files on cloud storage - the recommended parquet size for DeltaLake is 1GB so if I have 1TB table I have 1000 files.
2. the planning engine inside polars-rust gets more sophisticated and will leverage async downloading from multiple sources. I don't think that the Rust code should wait for python to download files on its behalf.

I have unfortunately been many cases in the situation where python code needs to be scaled up through multiprocessing and I am hoping that polars will allow me to delay that point in my daily work.

There are also use cases where polars-rust could be used as-is. The delta-rs library uses object_store, this is actually where I got the idea from. Doing the downloads on the rust side will enable pure-rust applications to leverage cloud rust.

And the last point is that moving enough functionality in rust will allow polars to be used from other languages, for example Node.

**Comment by chitralverma** [comment]
> > @winding-lines AFAIK, `scan_parquet` and `read_parquet` can rely on `fsspec` filesystems to read data directly from S3, GCS and many more. so do we really need a separate `read_parquet_cloud` for this?
> > 
> > For a library point of view, it will be great for the users if they were flawlessly able to access their cloud data sets from the existing methods ?
> 
> I personally believe that Python has achieved its amazing success in machine learning, data science and big data by providing great APIs on top of native libraries. You are right that fsspec is one such APIs, however I have concerns that it will impact performance in some use cases because of the way execution is serialized through Python. Some of these use cases:
> 
> 1. I need to open hundreds of parquet files on cloud storage - the recommended parquet size for DeltaLake is 1GB so if I have 1TB table I have 1000 files.
> 2. the planning engine inside polars-rust gets more sophisticated and will leverage async downloading from multiple sources. I don't think that the Rust code should wait for python to download files on its behalf.
> 
> I have unfortunately been many times in the situation where python code needs to be scaled up through multiprocessing and I am hoping that polars will allow me to delay that point in my daily work.
> 
> There are also use cases where polars-rust could be used as-is. The delta-rs library uses object_store, this is actually where I got the idea from. Doing the downloads on the rust side will enable pure-rust applications to leverage cloud storage.
> 
> And the last point is that moving enough functionality in rust will allow polars to be used from other languages, for example Node.

alright then, the extension to other languages definitely makes sense!

**Comment by ozgrakkurt** [comment]
closes #4865

**Review by ritchie46 — COMMENTED** [review]
Thank you for the PR @winding-lines. I think overall it looks good!

I have some remarks here and there.

File: polars/polars-io/src/parquet/mmap.rs [hunk 1]
```diff
@@ -6,24 +8,49 @@ use arrow::io::parquet::read::{
...
+/// 1. a local memory mapped file
+/// 2. data fetched from cloud storage on demand. The intent is to enable a two phase approach:
+///    a. identify all the needed columns
+///    b. asynchronously fetch them in parallel, for example using object_store
+///    c. store the data in this data structure
+///    d. when all the data is available deserialize on multiple threads, for example using rayon
+pub(crate) enum ColumnStore<'a> {
+    Local(&'a [u8]),
+    #[cfg(feature = "parquet-async")]
+    Fetched(&'a HashMap<u64, Vec<u8>>),
```
[the remark this answers is not in the document]
Comment (winding-lines) [reply polars/polars-io/src/parquet/mmap.rs:21]: Changed to the proper hashmap and improved documentation.

File: polars/polars-io/src/parquet/async_impl.rs [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+
+        Ok(ParquetObjectStore {
+            store,
+            path,
+            length: None,
+            metadata: None,
+        })
+    }
+
+    /// Initialize the lenght property of the object, unless it has already been fetched.
```
[the remark this answers is not in the document]
Comment (winding-lines) [reply polars/polars-io/src/parquet/async_impl.rs:46]: Ugh, thanks.

File: polars/polars-io/src/object_store.rs [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        buf: &mut [u8],
+    ) -> std::task::Poll<std::io::Result<usize>> {
+        // Use block_on in order to get the future result in this thread and copy the data in the output buffer.
+        // With this approach we keep ownership of the buffer and we don't have to pass it to the future runtime.
+        match block_on(self.read_operation(cx, buf.len())) {
+            Poll::Ready(Ok(bytes)) => {
+                buf.copy_from_slice(&bytes[..]);
```
[the remark this answers is not in the document]
Comment (winding-lines) [reply polars/polars-io/src/object_store.rs:105]: Good point, incrementing.

File: polars/polars-core/src/error.rs [hunk 4]
```diff
@@ -59,6 +59,8 @@ pub enum PolarsError {
     Io(#[from] std::io::Error),
     #[error("DuplicateError: {0}")]
     Duplicate(ErrString),
+    #[error("External: {0}")]
```
[the remark this answers is not in the document]
Comment (winding-lines) [reply polars/polars-core/src/error.rs:62]: Switched to anyhow, thanks!

File: polars/polars-io/Cargo.toml [hunk 5]
```diff
@@ -59,6 +64,8 @@ serde = { version = "1", features = ["derive"], optional = true }
 serde_json = { version = "1", optional = true, default-features = false, features = ["alloc", "raw_value"] }
 simd-json = { version = "0.7.0", optional = true, features = ["allow-non-simd", "known-key"] }
 simdutf8 = "0.1"
+tokio = { version = "1.22.0", features = ["full"], optional = true }
```
[the remark this answers is not in the document]
Comment (winding-lines) [reply polars/polars-io/Cargo.toml:67]: Done.

File: polars/polars-io/src/object_store.rs [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+type OptionalFuture = Arc<Mutex<Option<BoxFuture<'static, std::io::Result<Vec<u8>>>>>>;
+
+/// Adaptor to translate from AsyncSeek and AsyncRead to the object_store get_range API.
+pub struct CloudReader {
+    // The current position in the stream, it is set by seeking and updated by reading bytes.
+    pos: u64,
+    // The total size of the object is required when seeking from the end of the file.
+    length: Option<u64>,
+    // Hold an reference to the store in a thread safe way.
+    object_store: Arc<Mutex<Box<dyn ObjectStore>>>,
```
[the remark this answers is not in the document]
Comment (winding-lines) [reply polars/polars-io/src/object_store.rs:29]: This is a future aware Mutex, see import above. There doesn't seem to be a ReadWriteLock in futures.

File: polars/polars-io/src/parquet/async_impl.rs [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+                        .map(|(pos, data)| format!("({}, {})", pos, data.len()))
+                        .collect::<Vec<_>>()
+                        .join(",")
+                );
+            }
+
+            let store = mmap::ColumnStore::Fetched(&downloaded_per_filepos);
+
+            // Deserialize and build dataframes.
+            let dfs = match self.parallel {
```
[the remark this answers is not in the document]
Comment (winding-lines) [reply polars/polars-io/src/parquet/async_impl.rs:260]: Good idea, simplified the changes a lot.

**Comment by winding-lines** [comment]
Thanks for the feedback @ritchie46 , addressed most of it. The one exception is around the `Mutex`, this is from the `futures` library and the library doesn't seem to have a `ReadWriteLock`.

**Comment by ritchie46** [comment]
Thanks a lot @winding-lines! Great addition.

**Comment by winding-lines** [comment]
@ozgrakkurt can you test to see if this works for you?

**Comment by chitralverma** [comment]
@winding-lines @ritchie46 does this allow the following now?
- consistent globbed paths for scan* and read* methods
- passing array for paths for scan* and read* methods

**Comment by chitralverma** [comment]
@winding-lines @ritchie46 does this allow the following now?
- consistent globbed paths for scan* and read* methods
- passing array for paths for scan* and read* methods

**Comment by winding-lines** [comment]
This PR only allows one cloud url. I am happy to push this further but I don’t actually know all the use cases. Could you file an issue and describe how the rust API is supposed to work?

**Comment by ozgrakkurt** [comment]
> @ozgrakkurt can you test to see if this works for you?

I'll try to test when I get a chance

**Pull request merged**